File: go/500.600/519.RandomFlipMatrix.go
package main

import (
	"fmt"
	"math/rand"
)

type Solution struct {
	m, n, total int
	dp          map[int]int
}

func Constructor(m int, n int) Solution {
	return Solution{
		m:     m,
		n:     n,
		total: m * n,
		dp:    map[int]int{},
	}
}

func (t *Solution) Flip() (ans []int) {
	x := rand.Intn(t.total)
	t.total--
	if v, ok := t.dp[x]; ok {
		ans = []int{v / t.n, v % t.n}
	} else {
		ans = []int{x / t.n, x % t.n}
	}
	if v, ok := t.dp[t.total]; ok {
		t.dp[x] = v
	} else {
		t.dp[x] = t.total
	}
	return
}

func (t *Solution) Reset() {
	t.dp = map[int]int{}
	t.total = t.m * t.n
}

func main() {
	cls := Constructor(3, 1)
	fmt.Println(cls.Flip())
	fmt.Println(cls.Flip())
	fmt.Println(cls.Flip())
	cls.Reset()
	fmt.Println(cls.Flip())
}

File: 500.600/519.RandomFlipMatrix.cpp
#include <iostream>
#include <vector>
#include <ctime>
#include <cstdlib>
#include <unordered_map>
using namespace std;

class RandomFlipMatrix
{
public:
    RandomFlipMatrix(int m, int n) : m(m), n(n), total(m * n)
    {
        srand(time(nullptr));
    }

    vector<int> flip()
    {
        int x = rand() % total;
        vector<int> ans;
        --total;
        if (dp.contains(x))
        {
            ans = {dp[x] / n, dp[x] % n};
        }
        else
        {
            ans = {x / n, x % n};
        }
        // 将每次消耗掉的点，都移动到末尾
        if (dp.contains(total))
        {
            dp[x] = dp[total];
        }
        else
        {
            dp[x] = total;
        }
        return ans;
    }

    void reset()
    {
        total = m * n;
        dp.clear();
    }

private:
    int m{};
    int n{};
    int total{};
    unordered_map<int, int> dp;
};

void print(const vector<int> &ret)
{
    cout << "[" << ret[0] << " , " << ret[1] << "]" << endl;
}

int main()
{
    RandomFlipMatrix cls(3, 1);
    print(cls.flip());
    print(cls.flip());
    print(cls.flip());
    cls.reset();
    print(cls.flip());

    return 0;
}
